src/psyclone/tests/domain/lfric/builtins/X_innerproduct_X_test.py

```diff
@@ -0,0 +1,159 @@
+    returns the expected string and 2) we generate correct code
+    for the built-in operation which calculates inner product of
+    a real-valued field 'X' by itself as 'innprod = innprod +
+    X(:)*X(:)'. 3) Also test the 'metadata()' method.
+
+    '''
+    # Test metadata
+    metadata = LFRicXInnerproductXKern.metadata()
+    assert isinstance(metadata, LFRicKernelMetadata)
+    _, invoke_info = parse(
+        os.path.join(BASE_PATH,
+                     "15.9.2_X_innerproduct_X_builtin.f90"),
+        api=API)
+    psy = PSyFactory(API,
+                     distributed_memory=dist_mem).create(invoke_info)
+    # Test '__str__' method
+    first_invoke = psy.invokes.invoke_list[0]
+    kern = first_invoke.schedule.children[0].loop_body[0]
+    assert str(kern) == "Built-in: X_innerproduct_X (real-valued field)"
+    # Test code generation
+    code = str(psy.gen)
+
+    output = (
+        "      !\n"
+        "      ! Initialise field and/or operator proxies\n"
+        "      !\n"
+        "      f1_proxy = f1%get_proxy()\n"
+        "      f1_data => f1_proxy%data\n"
+        "      !\n")
+    assert output in code
+    if not dist_mem:
+        output_seq = (
+            "      ! Initialise number of DoFs for aspc1_f1\n"
+            "      !\n"
+            "      undf_aspc1_f1 = f1_proxy%vspace%get_undf()\n"
+            "      !\n"
+            "      ! Set-up all of the loop bounds\n"
+            "      !\n"
+            "      loop0_start = 1\n"
+            "      loop0_stop = undf_aspc1_f1\n"
+            "      !\n"
+            "      ! Call our kernels\n"
+            "      !\n"
+            "      !\n"
+            "      ! Zero summation variables\n"
+            "      !\n"
+            "      asum = 0.0_r_def\n"
+            "      !\n"
+            "      DO df=loop0_start,loop0_stop\n"
+            "        asum = asum + f1_data(df) * f1_data(df)\n"
+            "      END DO\n"
+            "      !\n")
+        assert output_seq in code
+    else:
+        output_dm = (
+            "      loop0_stop = f1_proxy%vspace%get_last_dof_owned()\n"
+            "      !\n"
+            "      ! Call kernels and communication routines\n"
+            "      !\n"
+            "      !\n"
+            "      ! Zero summation variables\n"
+            "      !\n"
+            "      asum = 0.0_r_def\n"
+            "      !\n"
+            "      DO df=loop0_start,loop0_stop\n"
+            "        asum = asum + f1_data(df) * f1_data(df)\n"
+            "      END DO\n"
+            "      global_sum%value = asum\n"
+            "      asum = global_sum%get_sum()\n"
+            "      !\n")
+        assert output_dm in code
+        assert "      USE scalar_mod, ONLY: scalar_type" in code
+        assert "      REAL(KIND=r_def), intent(out) :: asum\n" in code
+        assert "      TYPE(scalar_type) global_sum\n" in code
+
+    # Test compilation of generated code
+    assert LFRicBuild(tmpdir).code_compiles(psy)
+
+
```

Comment:
Is there any CodeCov missed lines if we remove all this?

This now should already be tested by the code lowering (below) and the backend (in another file), which are better encapsulated for unit testing.

If we mange to delete these it would be great because sometimes we want to change the Fortran formatting in the backend and at the moment we need to update these assert all over the tests, which is quite time consuming.

The same applies to all similar tests under src/psyclone/tests/domain/lfric/builtins/

Reply:
I removed the majority of this and left the key differences between serial code and distributed memory implementation. Similar for other tests mentioned above, with the exception of one test in `reduction_builtins_test.py` , where it seems to me that the intention is to showcase the entire PSy layer.